js/editor-ta.js

```diff
@@ -0,0 +1,304 @@
+  const calculatedWidth = this.mirror_.getBoundingClientRect().width;
+  // if the height is less then the window height just snap to the window height
+  if (calculatedHeight < this.initHeight_) {
+    this.container_.scrollTop = 0;
+    this.updateHeight(this.initHeight_);
+  } else {
+    this.updateHeight(calculatedHeight);
+  }
+
+  if (this.settings_.get('wraplines')) {
+    this.updateWidth(this.initWidth_);
+  } else {
+    this.updateWidth(Math.max(this.initWidth_, calculatedWidth));
+  }
+
+};
+
+
+EditorTextArea.prototype.updateLineNumbers = function() {
+  let linesRendered = this.lineNumbers_.children.length;
+
+  // remove excess elements
+  while (linesRendered > this.numLines_) {
+    this.lineNumbers_.lastChild.remove();
+    linesRendered--;
+  }
+  // add missing elements
+  while(linesRendered < this.numLines_) {
+    const line = linesRendered + 1;
+    // If word wrap is on we have to additionally set height
+    const height = this.settings_.get('wordwrap') ? this.getLineHeight(line) : undefined;
+    this.lineNumbers_.appendChild(this.lineNumberElem(line, height));
+    linesRendered++;
+  }
+
+};
+
+EditorTextArea.prototype.getLineHeight = function(lineNumber) {
+  // TODO(zafzal): Speed this up, splitting the whole doc is a bottleneck
+  const line = this.textarea_.value.split('\n')[lineNumber-1];
+  this.lineMirror_.innerText = line;
+  return this.lineMirror_.getBoundingClientRect().height;
+}
+
+EditorTextArea.prototype.undo = function() {
+  // This is handled by the text area defaults
+};
+
+EditorTextArea.prototype.redo = function() {
+  // This is handled by the text area defaults
+};
+
+EditorTextArea.prototype.focus = function() {
+  this.textarea_.focus();
+};
+
+/**
+ * @param {Session} session
+ * @param {string} extension
+ */
+EditorTextArea.prototype.setMode = function(session, extension) {
+  // textarea does not support any modes other then plain text :(
+};
+
+/**
+ * @param {number} fontSize
+ * Update font size from settings.
+ */
+EditorTextArea.prototype.setFontSize = function(fontSize) {
+  this.textarea_.style.fontSize = fontSize + 'px';
+  this.lineNumbers_.style.fontSize = fontSize + 'px';
+  this.mirror_.style.fontSize = fontSize + 'px';
+  this.lineMirror_.style.fontSize = fontSize + 'px';
+  this.updateLineNumbers();
+  this.updateTextArea();
+};
+
+/**
+ * @param {number} size
+ */
+EditorTextArea.prototype.setTabSize = function(size) {
+  // TODO(zafzal): this
+  // TODO(zafzal): think about how to indicate that a tab is not "tab" but "shift tab"
+};
+
+/**
+ * @param {string} theme
+ */
+EditorTextArea.prototype.setTheme = function(theme) {
+  // TODO(zafzal): this
+};
+
+/**
+ * @param {boolean} val
+ */
+EditorTextArea.prototype.showHideLineNumbers = function(val) {
+  // TODO(zafzal): this
+};
+
+/**
+ * @param {boolean} val
+ */
+EditorTextArea.prototype.setWrapLines = function(val) {
+  // just update the text area which will check the relevent settings item
+  this.updateTextArea();
```

Comment:
does it also need to do the height update that updateTextArea performs?

Reply:
Yep, the height of the document changes depending on if we are wrapping lines or not

Reply:
Ack.